agents-api/agents_api/queries/utils.py

```diff
@@ -0,0 +1,687 @@
+
+            not only_on_error and debug and pprint(
+                dict(
+                    result=result.to_dict(orient="records"),
+                )
+            )
+
+            return result
+
+        # Set the wrapped function as an attribute of the wrapper,
+        # forwards the __wrapped__ attribute if it exists.
+        setattr(wrapper, "__wrapped__", getattr(func, "__wrapped__", func))
+
+        return wrapper
+
+    if func is not None and callable(func):
+        return cozo_query_dec(func)
+
+    return cozo_query_dec
+
+
+def pg_query(
+    func: Callable[P, tuple[str | list[str | None], dict]] | None = None,
+    debug: bool | None = None,
+    only_on_error: bool = False,
+    timeit: bool = False,
+):
+    def pg_query_dec(func: Callable[P, tuple[str | list[Any], dict]]):
+        """
+        Decorator that wraps a function that takes arbitrary arguments, and
+        returns a (query string, variables) tuple.
+
+        The wrapped function should additionally take a client keyword argument
+        and then run the query using the client, returning a Record.
+        """
+
+        from pprint import pprint
+
+        from tenacity import (
+            retry,
+            retry_if_exception,
+            stop_after_attempt,
+            wait_exponential,
+        )
+
+        def is_resource_busy(e: Exception) -> bool:
+            return (
+                isinstance(e, HTTPException)
+                and e.status_code == 429
+                and not getattr(e, "cozo_offline", False)
+            )
+
+        @retry(
+            stop=stop_after_attempt(4),
+            wait=wait_exponential(multiplier=1, min=4, max=10),
+            retry=retry_if_exception(is_resource_busy),
+        )
+        @wraps(func)
+        async def wrapper(
+            *args: P.args, client=None, **kwargs: P.kwargs
+        ) -> list[Record]:
+            if inspect.iscoroutinefunction(func):
+                query, variables = await func(*args, **kwargs)
+            else:
+                query, variables = func(*args, **kwargs)
+
+            not only_on_error and debug and print(query)
+            not only_on_error and debug and pprint(
+                dict(
+                    variables=variables,
+                )
+            )
+
+            # Run the query
+            from ..clients import pg
+
+            try:
+                client = client or await pg.get_pg_client()
+
+                start = timeit and time.perf_counter()
+                results: list[Record] = await client.fetch(query, *variables)
+                end = timeit and time.perf_counter()
+
+                timeit and print(f"PostgreSQL query time: {end - start:.2f} seconds")
+
+            except Exception as e:
+                if only_on_error and debug:
+                    print(query)
+                    pprint(variables)
+
+                debug and print(repr(e))
+                connection_error = isinstance(
```

Comment:
are there errors in asyncpg that may be retryable? if so let's use them, if not, let's remove this block